[Merged by Bors] - feat(linear_algebra): quadratic forms

Vierkantor opened the pull request:
Define quadratic forms over a module, maps from quadratic forms to bilinear forms and matrices, positive definite quadratic forms and the discriminant of quadratic forms.

Along the way, I added some definitions to `data/matrix/basic.lean` and `linear_algebra/bilinear_form.lean` and did some cleaning up.

TO CONTRIBUTORS:

Please include a summary of the changes made in this PR above "TO CONTRIBUTORS:", as that text will become the commit message. You are also encouraged to append the following [co-authorship template](https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors) if you'd like to acknowledge suggestions / commits made by other users:

Co-authored-by: name <name@example.com>

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover-community/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover-community/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover-community/mathlib/blob/master/docs/contribute/doc.md)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover-community/mathlib/blob/master/docs/contribute/code-review.md)

If you're confused by comments on your PR like `bors r+` or `bors d+`, please see our [notes on bors](https://github.com/leanprover-community/mathlib/blob/master/docs/contribute/bors.md) for information on our merging workflow.

Vierkantor committed:
feat(linear_algebra): quadratic forms

The definition of a quadratic form over a module, maps to bilinear forms and
matrices, positive definite quadratic forms and discriminant of quadratic
forms.

jcommelin commented on src/linear_algebra/bilinear_form.lean:
```diff
@@ -40,8 +41,9 @@ structure bilin_form (R : Type u) (M : Type v) [ring R] [add_comm_group M] [modu
 (bilin_add_right : ∀ (x y z : M), bilin x (y + z) = bilin x y + bilin x z)
 (bilin_smul_right : ∀ (a : R) (x y : M), bilin x (a • y) = a * (bilin x y))
 
+/-- A linear map with two arguments is a bilinear form. -/
```
I feel like this could be confused with a map from `M × M`.
```suggestion
/-- A map with two arguments that is linear in both is a bilinear form -/
```

jcommelin replied:
Alternatively:
` A linear map to linear self-maps is a bilinear form`

jcommelin commented on src/linear_algebra/bilinear_form.lean:
```diff
@@ -125,9 +136,13 @@ instance to_module : module R₂ (bilin_form R₂ M) :=
 def to_linear_map : M →ₗ[R₂] M →ₗ[R₂] R₂ :=
 linear_map.mk₂ R₂ F.1 (bilin_add_left F) (bilin_smul_left F) (bilin_add_right F) (bilin_smul_right F)
 
+/-- Bilinear forms are equivalent to linear maps with two arguments. -/
```
Idem dito

jcommelin commented on src/linear_algebra/bilinear_form.lean:
```diff
@@ -136,8 +151,57 @@ def bilin_linear_map_equiv : (bilin_form R₂ M) ≃ₗ[R₂] (M →ₗ[R₂] M
+@[simp] lemma comp_right_comp_left (B : bilin_form R M) (l r : M →ₗ[R] M) :
+  (B.comp_right r).comp_left l = B.comp l r := rfl
+
+@[simp] lemma coe_fn_comp (B : bilin_form R N) (l r : M →ₗ[R] N) (v w) :
```
In other parts of the library we seem to name these `comp_apply`. Of course here we are applying two arguments... so it is slightly different. But I would still think the same naming convention could be used. What do you think?

Vierkantor replied:
The difference between "apply one argument" and "apply two arguments" doesn't seem too big, compared to "apply" and "coerce to a function". I'll change it :+1:

jcommelin commented on src/linear_algebra/bilinear_form.lean:
```diff
@@ -177,6 +241,102 @@ end
+    bilin_smul_left := λ a x y, by simp,
+    bilin_add_right := λ x y z, by simp [matrix.mul_add],
+    bilin_smul_right := λ a x y, by simp,
+  },
```
```suggestion
    bilin_smul_right := λ a x y, by simp, },
```

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+variables {R : Type u} {M : Type v} [add_comm_group M] [ring R]
+variables {R₁ : Type u} [comm_ring R₁] [module R₁ M]
+
+/-- Up to a factor 2, `Q.polar` is the associated bilinear form for a quadratic form `Q`. -/
```
I've never seen this terminology before. Out of curiosity, do you know where it comes from?

Vierkantor replied:
I saw this name in [one section of the WIkipedia article](https://en.wikipedia.org/wiki/Quadratic_form#Generalization). I'm not aware of a name for it in other sources (because they go directly to `associated`). If you have a more common name, I'd be happy to change it.

jcommelin replied:
Oooh, no, I didn't have a name either. To me it was always a nameless gadget. So I'm fine with this name. But it might be good to add a pointer to wiki in the docstring, because the name will not ring a bell with many people.

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+open quadratic_form
+/-- A quadratic form over a module. -/
+structure quadratic_form (R : Type u) (M : Type v) [ring R] [add_comm_group M] [module R M] :=
+(map : M → R)
```
In the rest of the library, the carrier function of a bundled map is usually called `to_fun`.
```suggestion
(to_fun : M → R)
```

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+/-- The `simp` normal form for a quadratic form is `coe_fn`, not `map`. -/
+@[simp] lemma map_eq_apply : Q.map = ⇑ Q := rfl
+
+lemma apply_smul (a : R) (x : M) : Q (a • x) = a * a * Q x := Q.map_smul a x
```
I think this should be called `map_smul`, to match what happens elsewhere. You could put a prime on the axioms in the structure to avoid a name clash.

Vierkantor replied:
I've renamed `quadratic_form.map` to `quadratic_form.to_fun` as you suggested. So the changes are `apply_smul -> map_smul` and `map_smul -> to_fun_smul`

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+
+instance : inhabited (quadratic_form R M) := ⟨0⟩
+
+instance : has_scalar R₁ (quadratic_form R₁ M) :=
```
What is the reason for the `R₁` as opposed to `R`?

Vierkantor replied:
I've added a notation section in the module docs to explain the distinction between `R` and `R₁`. In this specific instance: `R₁` is commutative, which is needed for `map_smul : Q (a * b * x) = b^2 * Q (a * x)`

jcommelin replied:
Ok, thanks!

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+  λ a x y, by simp [polar_to_quadratic_form, smul_left, smul_right, mul_add] ⟩
+
+@[simp] lemma to_quadratic_form_apply (B : bilin_form R M) :
+(B.to_quadratic_form : M → R) = λ x, B x x :=
```
Is there a reason why you write this as a lambda, instead of putting `x` as a hypothesis before the colon?

Vierkantor replied:
Not really. Fixed!

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+open bilin_form sym_bilin_form
+
+section associated
+variables [has_inv R₁] {B₁ : bilin_form R₁ M}
```
Do we to assume `has_inv`? Or should we ask explicitly for an inverse of `2`? Because with `has_inv` we cannot apply this to a general local ring (even though specific examples like `Z_[p]` do have a `has_inv` instance).

Vierkantor replied:
I wanted to use a typeclass because it fills in the argument automatically. It also expresses the intent that there is essentially a unique `associated` form. I've changed it into a bundled `{half : R // half * 2 = 1}` argument which at least solves the second part (compared to the unbundled version which only passes `half : R` to `associated` and adds `half * 2 = 1` in a few lemmas).

A random thought. How about defining:
```lean
class invertible [monoid R] (x : R) : Type := (inv : R) (inv_mul : inv * x = 1) (mul_inv : x * inv = 1)
```
with instances along the lines of `invertible 1`, `[division_ring R] [char_zero R] [invertible x] : invertible (bit0 x)`, `[division_ring R] [char_zero R] : invertible (bit1 x)`.

jcommelin replied:
Another option would be `[fact (is_unit 2)]`, along the lines of https://gist.github.com/cipher1024/9bd785a75384a4870b331714ec86ad6f#file-zmod_redesign-lean
On the other hand, that "fact" would not record data.

Vierkantor replied:
I think making sure the definitions stay computable is a good reason to take the inverse as data, so I would prefer this to `is_unit`. (I've actually tested the `discr` function by `#eval`ing it on examples.)

Vierkantor replied:
Updated to use `invertible` now that it got merged

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -0,0 +1,290 @@
+@[simp] lemma associated_apply (Q : quadratic_form R₁ M) (x y : M) :
+  Q.associated x y = 2⁻¹ * (Q (x + y) - Q x - Q y) := rfl
+
+lemma ssociated_is_sym (Q : quadratic_form R₁ M) : is_sym Q.associated :=
```
```suggestion
lemma associated_is_sym (Q : quadratic_form R₁ M) : is_sym Q.associated :=
```

jcommelin reviewed (commented):
Thanks for this PR! I'm excited... the classification of quadratic forms over `rat` is coming closer. I like this.

Vierkantor committed:
Fix linting errors

Vierkantor committed:
Apply suggestions from code review

Co-Authored-By: Johan Commelin <johan@commelin.net>

Vierkantor committed:
Apply suggestions from code review

Co-Authored-By: Johan Commelin <johan@commelin.net>

Vierkantor committed:
Apply suggestions from code review

Co-Authored-By: Johan Commelin <johan@commelin.net>

Vierkantor committed:
Pass around an inverse of `2` instead of a full `has_inv` instance

This generalizes the theory of quadratic forms to local rings that have an
inverse for 2 but not for all elements

Vierkantor committed:
Fix linting error

Vierkantor committed:
*really* fix linting errors

jcommelin reviewed (commented):
Apart from the `two_inv` decision, I'm really happy with this, and it's ready to merge.
(This doesn't mean that I don't like `two_inv`, just that I don't want to take that decision alone.)

Vierkantor committed:
Replace `inv_two` with `⅟2`
